pilot/pkg/xds/debuggen.go

```
@@ -0,0 +1,122 @@
+
+package xds
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	"github.com/golang/protobuf/ptypes/any"
+
+	"istio.io/istio/pilot/pkg/model"
+)
+
+const (
+	// TypeDebug requests debug info from istio, a secured implementation for istio debug interface
+	TypeDebug = "istio.io/debug"
+	debugURL  = "http://localhost:8080/debug/"
+)
+
+var activeNamespaceDebuggers = []string{
+	"config_dump",
+	"ndsz",
+	"syncz",
+	"edsz",
+	"config_distribution",
+}
+
+// DebugGen is a Generator for istio debug info
+type DebugGen struct {
+	Server *DiscoveryServer
+}
+
+func NewDebugGen(s *DiscoveryServer) *DebugGen {
+	return &DebugGen{
+		Server: s,
+	}
+}
+
+// Generate XDS debug responses according to the incoming debug request
+func (dg *DebugGen) Generate(proxy *model.Proxy, push *model.PushContext, w *model.WatchedResource, updates *model.PushRequest) (model.Resources, error) {
+	res := []*any.Any{}
+	var err error
+	var out []byte
+	if w.ResourceNames == nil {
+		res = append(res, &any.Any{
+			TypeUrl: TypeDebug,
+			Value:   []byte("debug type is required"),
+		})
+		return res, nil
+	}
+	if len(w.ResourceNames) != 1 {
+		return res, nil
+	}
+	resourceName := w.ResourceNames[0]
+	lastPos := strings.LastIndex(resourceName, "?")
+	firstPos := strings.Index(resourceName, "?")
+	systemNameSpace := fmt.Sprint(resourceName[lastPos+17:])
+	debugFullType := fmt.Sprint(resourceName[:lastPos])
+	identity := proxy.VerifiedIdentity
+	debugType := resourceName
+	if firstPos != -1 {
+		debugType = fmt.Sprint(resourceName[:firstPos])
+	}
+	if identity.Namespace != systemNameSpace {
+		shouldAllow := false
+		for _, debuggerName := range activeNamespaceDebuggers {
+			if debuggerName == debugType {
+				shouldAllow = true
+				break
+			}
+		}
+		if !shouldAllow {
+			res = append(res, &any.Any{
+				// nolint: staticcheck
+				TypeUrl: TypeDebug + "/" + debugFullType,
+				Value:   []byte(fmt.Sprintf("the debug info is not available for current identity: %q", identity)),
```

should return an error instead, not as a value I think

---

fixed

---

This is still returning it as a value